core/dbt/adapters/base/impl.py

```diff
@@ -1238,13 +1301,52 @@ def calculate_freshness(
 
         snapshotted_at = _utc(table[0][1], source, loaded_at_field)
         age = (snapshotted_at - max_loaded_at).total_seconds()
-        freshness = {
+        freshness: FreshnessResponse = {
             "max_loaded_at": max_loaded_at,
             "snapshotted_at": snapshotted_at,
             "age": age,
         }
         return adapter_response, freshness
 
+    def calculate_freshness_from_metadata(
+        self,
+        source: BaseRelation,
+        manifest: Optional[Manifest] = None,
+    ) -> Tuple[Optional[AdapterResponse], FreshnessResponse]:
+        kwargs: Dict[str, Any] = {
+            "information_schema": source.information_schema_only(),
+            "relations": [source],
+        }
+        result = self.execute_macro(
+            GET_RELATION_LAST_MODIFIED_MACRO_NAME, kwargs=kwargs, manifest=manifest
+        )
+        adapter_response, table = result.response, result.table  # type: ignore[attr-defined]
+
+        try:
+            assert len(table) == 1  # There should be one row since we requested one relation
```

Comment:
I just learned this recently, `assert` does nothing in functional code. The test runner flips a switch under the hood to make it do something, but you would never want that switch on in production. If you want to catch an edge case here, then you need to use a regular if block.

Reply:
Got rid of this assertion.